### PR TITLE
Add ledger support for NearProver upgrade

### PR DESCRIPTION
**Pull request by karim-en**

### CODE REVIEW

File: contracts/eth/nearprover/utils/eth-ledger-signer.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        setTimeout(resolve, duration);
+    });
+}
+class EthLedgerSigner extends ethers.Signer {
+    constructor(provider, path) {
+        super();
+        if (!path) {
+            path = defaultPath;
+            console.log(defaultPath)
+        }
```
Comment (sept-en) [inline contracts/eth/nearprover/utils/eth-ledger-signer.js:31]:
```suggestion
        if (!path) {
            path = defaultPath;
            console.log('Using default HD path');
        }
        
        console.log(`Using Ledger key path: ${path}`);
```

File: contracts/eth/nearprover/hardhat.config.js [hunk 2]
```diff
@@ -9,9 +9,10 @@ const INFURA_API_KEY = process.env.INFURA_API_KEY || '';
 task('upgrade-provers-bridge-address-to', 'Upgrades the provided prover to use the bridge at the provided address')
     .addParam('prover', 'Prover address')
     .addParam('newBridge', 'The address of the new bridge')
+    .addParam('ledgerKey', 'The ledger key path to sign transactions', undefined, undefined, true)
```
Comment (sept-en) [inline contracts/eth/nearprover/hardhat.config.js:12]: It's actually not "Ledger key", but rather "Ledger key path".
Please change the parameter name accordingly

File: contracts/eth/nearprover/utils/eth-ledger-signer.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            return eth.getAppConfiguration().then((_config) => {
+                return eth;
+            }, (error) => {
+                return Promise.reject(error);
+            });
+        }, (error) => {
+            return Promise.reject(error);
+        }));
+    }
+    _retry(callback, timeout) {
```
Comment (sept-en) [inline contracts/eth/nearprover/utils/eth-ledger-signer.js:45]: Add an empty line here and in other places in the file. Even if this formatting was used in the original repo, it's worth fixing.

File: contracts/eth/nearprover/utils/eth-ledger-signer.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    async signTransaction(transaction) {
+        const tx = await ethers.utils.resolveProperties(transaction);
+        const baseTx = {
+            chainId: tx.chainId || undefined,
+            data: tx.data || undefined,
+            gasLimit: tx.gasLimit || undefined,
+            gasPrice: tx.gasPrice || undefined,
+            nonce: tx.nonce ? ethers.BigNumber.from(tx.nonce).toNumber() : undefined,
+            maxFeePerGas: tx.maxFeePerGas || undefined,
+            maxPriorityFeePerGas: tx.maxPriorityFeePerGas || undefined,
```
Comment (sept-en) [inline contracts/eth/nearprover/utils/eth-ledger-signer.js:92]: Please add the comment which will explain that these lines has been added as they haven't been yet accepted in the original repo, providing an appropriate link to PR

File: contracts/eth/nearprover/hardhat.config.js (hunk 2 above)
Comment (sept-en) [reply contracts/eth/nearprover/hardhat.config.js:12]: Also, please add a separate `make` command that uses this option

**Review by sept-en — APPROVED** [review]
I added some minor comments, but other than that LGTM!

File: contracts/eth/nearprover/utils/eth-ledger-signer.js (hunk 1 above)
Comment (karim-en) [reply contracts/eth/nearprover/utils/eth-ledger-signer.js]: Fixed at e5f44756a4da2b1b886ed1e94421532ed5fa1d42

File: contracts/eth/nearprover/utils/eth-ledger-signer.js (hunk 3 above)
Comment (karim-en) [reply contracts/eth/nearprover/utils/eth-ledger-signer.js]: fixed at e5f44756a4da2b1b886ed1e94421532ed5fa1d42

File: contracts/eth/nearprover/utils/eth-ledger-signer.js (hunk 4 above)
Comment (karim-en) [reply contracts/eth/nearprover/utils/eth-ledger-signer.js]: fixed at e5f44756a4da2b1b886ed1e94421532ed5fa1d42

File: contracts/eth/nearprover/UPGRADE.md [hunk 5]
```diff
@@ -26,6 +26,8 @@ To deploy the prover with the mock run: <br/>
 
 2. Upgrade the bridge address for the provided prover:<br/>
 `$ make upgrade-provers-bridge-address-to PROVER=<PROVER_ADDRESS_HERE> NEW_BRIDGE=<BRIDGE_ADDRESS_HERE>`
+or
+`$ make upgrade-provers-bridge-address-to-with-ledger PROVER=<PROVER_ADDRESS_HERE> NEW_BRIDGE=<BRIDGE_ADDRESS_HERE> LEDGER_KEY_PATH=[KEY_PATH]`
```
Comment (sept-en) [inline contracts/eth/nearprover/UPGRADE.md:30]:
```suggestion
`$ make upgrade-provers-bridge-address-to-with-ledger PROVER=<PROVER_ADDRESS_HERE> NEW_BRIDGE=<BRIDGE_ADDRESS_HERE> LEDGER_KEY_PATH=<LEDGER_KEY_PATH_HERE>`
```

File: contracts/eth/nearprover/utils/upgrade_bridge_address.js [hunk 6]
```diff
@@ -15,21 +16,28 @@ async function upgradeProversBridgeAddressTo(provider, proverAddress, newBridgeA
...
+    const mask = ethers.BigNumber.from('0x000000000000000000000000ffffffffffffffffffffffffffffffffffffffff');
     const response = await nearProver
-        .connect(adminWallet)
+        .connect(signer)
         .adminSstoreWithMask(BRIDGE_ADDRESS_SLOT, newBridgeAddress, mask);
-    await response.wait(10).then(function(receipt) {
-        console.log(`Transaction mined: `);
+    console.log(response);
+    console.log('Waiting for tx confirmation...');
+    await response.wait(10).then(function (receipt) {
```
Comment (0x3bfc) [inline contracts/eth/nearprover/utils/upgrade_bridge_address.js:39]: Is the `10` seconds enough to get the tx confirmation ?

Comment (karim-en) [reply contracts/eth/nearprover/utils/upgrade_bridge_address.js:39]: @0x3bfc it is a number of confirms not seconds [ref](https://docs.ethers.io/v5/api/providers/types/#providers-TransactionResponse)

File: contracts/eth/nearprover/utils/upgrade_bridge_address.js [hunk 7]
```diff
@@ -15,21 +16,28 @@ async function upgradeProversBridgeAddressTo(provider, proverAddress, newBridgeA
     console.log(`Initial bridge address: ${initialBridgeAddress}`);
     console.log(`Trying to upgrade bridge address to: ${newBridgeAddress}`);
 
-    const adminWallet = new ethers.Wallet(process.env.ETH_PRIVATE_KEY, provider);
+    let signer;
+    if (ledgerKeyPath != null) {
```
Comment (mfornet) [inline contracts/eth/nearprover/utils/upgrade_bridge_address.js:20]:
```suggestion
    if (ledgerKeyPath !== null) {
```

File: contracts/eth/nearprover/UPGRADE.md [hunk 8]
```diff
@@ -26,6 +26,8 @@ To deploy the prover with the mock run: <br/>
 
 2. Upgrade the bridge address for the provided prover:<br/>
 `$ make upgrade-provers-bridge-address-to PROVER=<PROVER_ADDRESS_HERE> NEW_BRIDGE=<BRIDGE_ADDRESS_HERE>`
+or
+`$ make upgrade-provers-bridge-address-to-with-ledger PROVER=<PROVER_ADDRESS_HERE> NEW_BRIDGE=<BRIDGE_ADDRESS_HERE> LEDGER_KEY_PATH=<LEDGER_KEY_PATH_HERE>`
```
Comment (mfornet) [inline contracts/eth/nearprover/UPGRADE.md:30]: Why in step 1 it says `eth-custodian` rather than `nearprover`?

Comment (karim-en) [reply contracts/eth/nearprover/UPGRADE.md:30]: It is a typo and it is already fixed by @sept-en at 0b128d9c3a06aac2dc115182e4afaf36861f541d

File: contracts/eth/nearprover/utils/upgrade_bridge_address.js (hunk 7 above)
Comment (karim-en) [reply contracts/eth/nearprover/utils/upgrade_bridge_address.js]: fixed at a47a53fc300ad83bc5b04418fccaac2052e1c60b

**Review by 0x3bfc — APPROVED** [review]
LGTM

File: contracts/eth/nearprover/utils/eth-ledger-signer.js [hunk 9]
```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2019 Richard Moore
```
[the remark this answers is not in the document]
Comment (karim-en) [reply contracts/eth/nearprover/utils/eth-ledger-signer.js:1]: done https://github.com/aurora-is-near/rainbow-bridge/issues/653

**Review by mfornet — APPROVED** [review]

**Pull request merged**